File: hub/src/light/effect/translation_effects/sparkle.py
import random

from light.effect.partial_effect import PartialEffect
from light.led_selector import LEDSelector
from light.effect.resolved_effect import ResolvedEffect
from light.effect.effect_priority import EffectPriority
from util.scale_functions import ScaleFunctions

class Sparkle(PartialEffect):
    def __init__(self, targets, startingDensity, sparkleSize, numIterations, duration):
        super().__init__()
        self.targets = targets
        self.startingDensity = startingDensity
        self.sparkleSize = sparkleSize
        self.duration = duration

    def getEffect(self, t):
        pos = []
        for target in self.targets:
            tempPos = []
            tNorm = 1 - ScaleFunctions.linear(t, self.startTime, self.duration)
            targetPos = LEDSelector().getAllTargetPositions(target)
            numLeds = len(targetPos)
            
            scaledDensity = int(self.startingDensity * tNorm)
            for sparkle in range(0, scaledDensity):
                randomLed = LEDSelector().unalias(target, random.randint(-1, numLeds))
                tempPos.append(randomLed)

            pos.extend(tempPos)

        return ResolvedEffect.createTranslation(list(set(tempPos)))

    def isTemporal(self):
        # Constant targets do not control timing
        return True

    def isComplete(self, t):
        return t > self.duration

File: hub/drumlite-js/src/light/effect/positional/Sparkle.ts
import { EffectTarget } from "../EffectTarget"
import PartialEffect, { EffectParameters, EffectParameter } from "../PartialEffect"
import ScaleFunctions from "../../../util/ScaleFunctions";
import LEDSelector from "../../LEDSelector";
import Util from "../../../util/Util";
import ResolvedEffect from "../../../effect/ResolvedEffect";

export class SparkleParams extends EffectParameters {
    effectName = "Sparkle";
    typeName = "Positional";

    constructor(targets: EffectTarget[] = [], density: number = 1, sparkleSize: number = 1, duration: number = 0) {
        super(0);
        this.params.targets = new EffectParameter<EffectTarget[]>("Targets", targets, "target", true)
        this.params.density = new EffectParameter<number>("Density", density)
        this.params.sparkleSize = new EffectParameter<number>("Sparkle Size", sparkleSize)
        this.params.duration = new EffectParameter<number>("Duration", duration)
    }
}

export default class Sparkle extends PartialEffect<SparkleParams> {
    constructor(params: SparkleParams, dt = 0) {
        super(params, dt);
    }

    public getEffect(t: number) {
        const ledSelector = new LEDSelector();
        const pos: number[] = [];
        for (let target of this.params.params.targets.val) {
            const tempPos: number[] = [];
            const tNorm = 1 - ScaleFunctions.linear(t, this.params.params.startTime.val, this.params.params.duration.val);
            const targetPos = ledSelector.getAllTargetPositions(target);
            const numLeds = targetPos.length;
            
            const scaledDensity = Math.floor(this.params.params.density.val * tNorm);
            for (let sparkle of Util.range(0, scaledDensity)) {
                const randomLed = ledSelector.unalias(target, Math.round(Math.random() * (numLeds - 1)));
                tempPos.push(randomLed);
            }

            pos.push(...tempPos);
        }

        return ResolvedEffect.createTranslation([...new Set<number>(pos)]);
    }

    public isTemporal() {
        // Constant targets do not control timing
        return true;
    }

    public isComplete(t: number) {
        return t > this.params.params.duration.val;
    }
}
